src/crl/mod.rs

```diff
@@ -0,0 +1,825 @@
+    if crl.issuer() != path.cert.issuer() {
+        return false;
+    }
+
+    let crl_idp = match (
+        path.cert.crl_distribution_points(),
+        crl.issuing_distribution_point(),
+    ) {
+        // If the certificate has no CRL distribution points, and the CRL has no issuing distribution point,
+        // then we can consider this CRL authoritative based on the issuer matching.
+        (cert_dps, None) => return cert_dps.is_none(),
+
+        // If the CRL has an issuing distribution point, parse it so we can consider its scope
+        // and compare against the cert CRL distribution points, if present.
+        (_, Some(crl_idp)) => {
+            match IssuingDistributionPoint::from_der(untrusted::Input::from(crl_idp)) {
+                Ok(crl_idp) => crl_idp,
+                Err(_) => return false, // Note: shouldn't happen - we verify IDP at CRL-load.
+            }
+        }
+    };
+
+    crl_idp.authoritative_for(path)
+}
+
+// When verifying CRL signed data we want to disambiguate the context of possible errors by mapping
+// them to CRL specific variants that a consumer can use to tell the issue was with the CRL's
+// signature, not a certificate.
+fn crl_signature_err(err: Error) -> Error {
+    match err {
+        Error::UnsupportedSignatureAlgorithm => Error::UnsupportedCrlSignatureAlgorithm,
+        Error::UnsupportedSignatureAlgorithmForPublicKey => {
+            Error::UnsupportedCrlSignatureAlgorithmForPublicKey
+        }
+        Error::InvalidSignatureForPublicKey => Error::InvalidCrlSignatureForPublicKey,
+        _ => err,
+    }
+}
+
+/// Describes how much of a certificate chain is checked for revocation status.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum RevocationCheckDepth {
+    /// Only check the end entity (leaf) certificate's revocation status.
+    EndEntity,
+    /// Check the revocation status of the end entity (leaf) and all intermediates.
+    Chain,
+}
+
+/// Describes how to handle the case where a certificate's revocation status is unknown.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum UnknownStatusPolicy {
+    /// Treat unknown revocation status permissively, acting as if the certificate were
+    /// not revoked.
+    Allow,
+    /// Treat unknown revocation status as an error condition, yielding
+    /// [Error::UnknownRevocationStatus].
+    Deny,
+}
+
+// Zero-sized marker type representing positive assertion that revocation status was checked
+// for a certificate and the result was that the certificate is not revoked.
+pub(crate) struct CertNotRevoked(());
+
+impl CertNotRevoked {
+    // Construct a CertNotRevoked marker.
+    fn assertion() -> Self {
+        Self(())
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+/// An opaque error indicating the caller must provide at least one CRL when building a
+/// [RevocationOptions] instance.
+pub struct CrlsRequired(pub(crate) ());
+
+pub(crate) struct IssuingDistributionPoint<'a> {
```

Comment:
I think I'm generally on-board with the `crl.rs` split you have in this branch, but it feels like `IssuingDistributionPoint` is a candidate to live in `types.rs`.

WDYT? Was there an explicit reason you thought otherwise?

Reply:
I guess my main reason was that keeping it in `mod.rs` keeps the two files more balanced in terms of size which seems nice? But I guess you're right that in terms of coupling, `IssuingDistributionPoint` really belongs in `types`. The other reason I could see is that I feel like the logic in `IssuingDistributionPoint::authoritative_for()` really belongs in `mod.rs`...

Added in an extra commit now, what do you think? (I could squash the commits if we want to go this way.)

Reply:
> The other reason I could see is that I feel like the logic in IssuingDistributionPoint::authoritative_for() really belongs in mod.rs...

Hm I was thinking the opposite and that maybe the free-standing `crl_authoritative` fn from `mod.rs` should be in `types.rs` associated with `&dyn CertRevocationList`. It's the one and only caller of `IssuingDistributionPoint::authoritative_for`.

It seems like maybe we could do something like this in `types.rs`:
```
impl dyn CertRevocationList {
    fn authoritative_for(&self, path: &PathNode<'_>) -> bool {
      ...
    }
}
```

Would it be unusual to add an `impl` for a `dyn` Trait like that? :thinking: I didn't want that to be part of the trait interface itself as a default fn since it implies you might want to provide a different implementation for the types that implement the trait and that's not the case.

Reply:
> Added in an extra commit now, what do you think? (I could squash the commits if we want to go this way.)

I'm a fan either way :+1:

Reply:
Squashed them.